### PR TITLE
Poincare projection

### PR DESCRIPTION
**Pull request by ferrine**
add flexible choice for projection epsilon in Poincare Ball model

### CODE REVIEW

File: geoopt/manifolds/poincare/math.py [hunk 1]
```diff
@@ -509,7 +513,7 @@ def _dist0(x, c, keepdim: bool = False, dim: int = -1):
     return dist_c * 2 / sqrt_c
 
 
-def clip_tangent(x, u, *, c=1.0, dim=-1):
+def clip_tangent(x, u, *, c=1.0, dim=-1, eps=None):
```
Comment (newkozlukov) [inline geoopt/manifolds/poincare/math.py]: `eps` could be smth interpretable, e.g. "result is such that adding a parallel vector of length `L` increases absolute velocity at least by `f(L, eps)`

File: geoopt/manifolds/poincare/math.py [hunk 2]
```diff
@@ -528,26 +532,26 @@ def clip_tangent(x, u, *, c=1.0, dim=-1):
...
 
-def _clip_tangent(x, u, c, dim: int = -1):
+def _clip_tangent(x, u, c, dim: int = -1, eps=None):
     # get the almost infinite vecotor estimate
     # this is the norm of travel vector to the opposite pole
     s = x.size(dim)
     p = torch.ones((s,), dtype=x.dtype, device=x.device)
     p = p / s ** 0.5 / (c ** 0.5)
-    p = _project(p, c, dim=dim)
+    p = _project(p, c, eps=eps)
```
Comment (newkozlukov) [inline geoopt/manifolds/poincare/math.py]: You sure you wanted to remove `dim=dim`?

**Comment by ferrine** [comment]
Not sure

сб, 27 апр. 2019 г., 15:35 Sergei Kozlukov <notifications@github.com>:

File: geoopt/manifolds/poincare/math.py (hunk 2 above)
Comment (ferrine) [reply geoopt/manifolds/poincare/math.py]: This is intended here

File: geoopt/manifolds/poincare/math.py (hunk 1 above)
Comment (ferrine) [reply geoopt/manifolds/poincare/math.py]: I removed this useless function

File: geoopt/manifolds/poincare/math.py [hunk 3]
```diff
@@ -84,7 +84,7 @@ def _eps(x):
 
 
 def _project(x, c, dim: int = -1, eps: float = None):
-    norm = x.norm(dim=dim, keepdim=True, p=2)
+    norm = x.norm(dim=dim, keepdim=True, p=2).clamp_min(1e-15)
```
Comment (newkozlukov) [inline geoopt/manifolds/poincare/math.py]: Please introduce some sort of `EPS = 1e-15` even if it be package-level global constant

File: geoopt/manifolds/poincare/math.py [hunk 4]
```diff
@@ -343,13 +343,12 @@ def mobius_coadd(x, y, *, c=1.0, dim=-1):
...
 
 def _mobius_coadd(x, y, c, dim: int = -1):
-    y = y + 1e-15
     x2 = x.pow(2).sum(dim=dim, keepdim=True)
     y2 = y.pow(2).sum(dim=dim, keepdim=True)
     num = (1 - c * y2) * x + (1 - c * x2) * y
     denom = 1 - c ** 2 * x2 * y2
     # avoid division by zero in this way
-    return num / (denom + 1e-15)
+    return num / denom.clamp_min(1e-15)
```
Comment (newkozlukov) [inline geoopt/manifolds/poincare/math.py]: Good. Now one at least doesn't have to doubt whether division by zero might appear because of negative `denom`. And hard threshold should probably produce less artifacts than soft threshold... in intermediate computations.

Still Mobius addition could in theory be generalized to velocities outside the admissible ball (TODO: reference), so I would've probably allowed negative `denom` with hard threshold reducing its absolute value but not changing the sign

File: geoopt/manifolds/poincare/math.py [hunk 5]
```diff
@@ -826,7 +775,7 @@ def logmap0(y, *, c=1.0, dim=-1):
 def _logmap0(y, c, dim: int = -1):
     sqrt_c = c ** 0.5
     y = y + 1e-15
```
Comment (newkozlukov) [inline geoopt/manifolds/poincare/math.py]: `clamp_min`

File: geoopt/manifolds/poincare/math.py [hunk 6]
```diff
@@ -1150,11 +1097,11 @@ def dist2plane(x, p, a, *, c=1.0, keepdim=False, signed=False, dim=-1):
...
-    diff_norm2 = diff.pow(2).sum(dim=dim, keepdim=keepdim)
+    diff_norm2 = diff.pow(2).sum(dim=dim, keepdim=keepdim).clamp_min(1e-15)
     sc_diff_a = (diff * a).sum(dim=dim, keepdim=keepdim)
     if not signed:
         sc_diff_a = sc_diff_a.abs()
-    a_norm = a.norm(dim=dim, keepdim=keepdim, p=2)
+    a_norm = a.norm(dim=dim, keepdim=keepdim, p=2).clamp_min(1e-15)
     num = 2 * sqrt_c * sc_diff_a
     denom = (1 - c * diff_norm2) * a_norm
     return arsinh(num / (denom + 1e-15)) / sqrt_c
```
Comment (newkozlukov) [inline geoopt/manifolds/poincare/math.py]: Same

**Comment by ferrine** [comment]
Sure, thanks for review

сб, 4 мая 2019 г., 17:06 Sergei Kozlukov <notifications@github.com>:

**Comment by ferrine** [comment]
@newkozlukov do you have any other concerns about this PR?

File: geoopt/manifolds/poincare/math.py (hunk 4 above)
Comment (ferrine) [reply geoopt/manifolds/poincare/math.py]: Not sure if we intend to provide the generalized version. Moreover, having admissible velosities I was able to achieve small negative denom due to numerical precision

File: geoopt/manifolds/poincare/math.py (hunk 5 above)
Comment (ferrine) [reply geoopt/manifolds/poincare/math.py]: Done

File: geoopt/manifolds/poincare/math.py (hunk 6 above)
Comment (ferrine) [reply geoopt/manifolds/poincare/math.py]: Done
